app/models/fact_value.rb

```diff
@@ -81,13 +81,15 @@ def self.count_each(fact, options = {})
     output
   end
 
-  def self.build_facts_hash(facts)
-    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
-
+  def self.build_facts_hash(facts, single_host = false)
     hash = {}
+    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
```

Comment:
We can improve performance quite a lot if we don't load fact value objects to memory at all:
``` ruby
hosts = Host.where(:id => facts.select(:host_id).uniq).all
```